### PR TITLE
openHAB JAX-RS publisher

### PR DESCRIPTION
**Pull request by maggu2810**
Do not use the `com.eclipsesource.jaxrs` `publisher` anymore.
It's service tracker breaks our OSGi runtime a little bit.

There has been already:
https://github.com/eclipse/smarthome/pull/4153

This PR uses DS instead of a service tracker, this ensure the consumed services are already ready.

It drops the option to add a whitelist as fragment of the io.rest.optimize bundle, but I am pretty sure this is not used in openHAB anyhwere (IIRC it has been added to allow third party products to use a different whitelist, but I was not aware of anyone ever used it).

Perhaps it will also help us with a more step by step transition to the JAX-RS Whiteboard pattern as a positive side effect. If not, it does not harm.

The used EclipseSource Provider sources already uses EPL but v1.
I kept the origin license headers and copied a notice file from another bundle.
Perhaps this needs be fixed.

Using that publisher I do not see any CNFE or other exceptions in the demo application anymore.

### CODE REVIEW

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/thirdparty/com/eclipsesource/jaxrs/publisher/ApplicationConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * Copyright (c) 2015 EclipseSource and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Holger Staudacher - initial API and implementation, ongoing development
+ ******************************************************************************/
+package org.openhab.core.thirdparty.com.eclipsesource.jaxrs.publisher;
```
Comment (kaikreuzer) [inline bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/thirdparty/com/eclipsesource/jaxrs/publisher/ApplicationConfiguration.java]: What we usually did (within bindings that required to copy 3rd party source code) is to keep the original namespace in place (i.e. `com.eclipsesource.jaxrs.publisher`) and just move those sources out from `src/main/java` to `src/3rdparty/java`. This also makes sure that an automatic license header update can nicely exclude such files. Would that be an option here as well?

Comment (maggu2810) [reply bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/thirdparty/com/eclipsesource/jaxrs/publisher/ApplicationConfiguration.java]: I changes the original namespace to ensure it is not used anymore (e.g. the rest.auth package used a reference to a provider class).
I don't know about all the other jersey stuff if it uses optional imports etc. And I don't want to inspect them all (better drop them some time).

I also modified and removed (only a few ATM) classes. So it is not about bundling unmodified stuff.

But sure, we could use `src/3rdparty/main` if we also "know" that it does not contain the originap sources anymore.

I do not have access to a PC this weekend, can you apply the requested changes?

Comment (kaikreuzer) [reply bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/thirdparty/com/eclipsesource/jaxrs/publisher/ApplicationConfiguration.java]: Ok, let's keep the namespace and just move the classes to `src/3rdparty/main`. I'll create a PR against your branch, sure!

Comment (maggu2810) [reply bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/thirdparty/com/eclipsesource/jaxrs/publisher/ApplicationConfiguration.java]: Thank you.

You are absolutely riggt, already the license header makes this necessary.

Don't forget to "inform" Maven and Eclipse to add that source folder :wink:

The mobile phonr will be enough to accept your PR.

File: bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java [hunk 2]
```diff
@@ -74,8 +74,8 @@ public static Response createErrorResponse(Response.StatusType status, String er
      * Depending on the status, create a Response object containing either the entity alone or an error JSON
      * which might hold the entity as well.
      *
-     * @param status the status for the response.
-     * @param entity the entity which is transformed into a JSON stream.
+     * @param status       the status for the response.
```
Comment (cweitkamp) [inline bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java]: Some minor formatting stuff. Maybe this https://github.com/eclipse/smarthome/issues/6424#issuecomment-471728524 helps.

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    public void addOhRESTResource(final ServiceReference<RESTResource> resource) {
+        jaxRsConnector.addResource(resource);
+    }
+
+    public void removeOhRESTResource(final ServiceReference<RESTResource> resource) {
+        jaxRsConnector.removeResource(resource);
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    public void addOhSSEResource(final ServiceReference<SseResource> resource) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java]: addOhSseResource

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        jaxRsConnector.addResource(resource);
+    }
+
+    public void removeOhSSEResource(final ServiceReference<SseResource> resource) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java]: removeOhSseResource

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/JAXRSConnectorProvider.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.core.thirdparty.com.eclipsesource.jaxrs.publisher.internal.JAXRSConnector;
+
+/**
+ * A provider of an JAX RS connector.
+ *
+ * @author Markus Rathgeb - Initial contribution
+ */
+@NonNullByDefault
+public interface JAXRSConnectorProvider {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/JAXRSConnectorProvider.java]: You have chosen `OhJaxRsResource` for the other interface, but `JAXRSConnectorProvider` here. Can we reach an agreement about the casing?

File: bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/OhJaxRsResource.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Marker interface for an openHAB JAX RS resource that needs to be registered by the publisher.
+ *
+ * <p>
+ * This marker is for providing OSGi services for services that should be registered by the REST publisher.
+ * You should not use this. It is only for a migration path...
+ *
+ * @author Markus Rathgeb - Initial contribution
+ */
+public interface OhJaxRsResource {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/OhJaxRsResource.java]: Add `@NonNullByDefault` annotation?

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/JAXRSConnectorProvider.java (hunk 5 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/JAXRSConnectorProvider.java]: You are absolutely right. We need to be consistent.
We are using most of the case upper case if the official abbreviation does the same (e.g. DTO and IP), there are also one exception you pointed below (Sse) but I think upper case maies sense.

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java (hunk 3 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java]: We should the same notation as se variable, so I agree "Sse" here.

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java (hunk 4 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java]: see above

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java (hunk 3 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java]: Okay, as we also prefixed by Oh, perhaps we can stay at SSE, to use the more common style in openHAB code base.
WDYT?

File: bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java (hunk 2 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java]: This has been done by the official IDE, we should check that setup (and yes, it should be fixed here, too)

File: bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/OhJaxRsResource.java (hunk 6 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/OhJaxRsResource.java]: Yes, sure as for every new class

**Comment by maggu2810** [comment]
@cweitkamp As also written to Kai above, I cannot access a PC, only my mobile phone. Would you like to create a PR for my branch?

**Comment by cweitkamp** [comment]
> Would you like to create a PR for my branch?

Yes, I can do that.

File: bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java (hunk 2 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java]: Done

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java (hunk 3 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/ResourceRegistrator.java]: I decided do keep SSE to be consistent to the other naming (regardless that the class itself does not fit to the naming scheme -- another story).
If you don't feel comfortable about, please comment again...

File: bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/JAXRSConnectorProvider.java (hunk 5 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest.publisher/src/main/java/org/openhab/core/io/rest/publisher/internal/JAXRSConnectorProvider.java]: Changes `OhJaxRsResource` to `OhJAXRSResource`.

File: bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/OhJaxRsResource.java (hunk 6 above)
Comment (maggu2810) [reply bundles/org.openhab.core.io.rest/src/main/java/org/eclipse/smarthome/io/rest/OhJaxRsResource.java]: Done (and marked deprecated as it should not be used for new stuff at all).
As written in the JavaDoc it is only a migration path.

**Comment by maggu2810** [comment]
IMHO I addressed the review comments of @cweitkamp

**Comment by cweitkamp** [comment]
I got some NPEs now during shutdown. Probably related to this PR. Any ideas?

```
2019-07-03 09:20:21.586 [ERROR] [org.openhab.core.io.rest.publisher  ] - bundle org.openhab.core.io.rest.publisher:2.5.0.201907030319 (157)[org.openhab.core.io.rest.publisher.internal.ResourceRegistrator(213)] : The removeOhJaxRsResource method has thrown an exception
java.lang.NullPointerException: null
	at org.openhab.core.thirdparty.com.eclipsesource.jaxrs.publisher.internal.JAXRSConnector.getPort(JAXRSConnector.java:175) ~[?:?]
	at org.openhab.core.thirdparty.com.eclipsesource.jaxrs.publisher.internal.JAXRSConnector.doRemoveResource(JAXRSConnector.java:220) ~[?:?]
	at org.openhab.core.thirdparty.com.eclipsesource.jaxrs.publisher.internal.JAXRSConnector.removeResource(JAXRSConnector.java:213) ~[?:?]
	at org.openhab.core.io.rest.publisher.internal.ResourceRegistrator.removeOhJaxRsResource(ResourceRegistrator.java:90) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:228) ~[42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41) ~[42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:664) ~[42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:510) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.inject.methods.BindMethod.invoke(BindMethod.java:42) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.DependencyManager.invokeUnbindMethod(DependencyManager.java:1927) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.DependencyManager.close(DependencyManager.java:1682) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.disposeImplementationObject(SingleComponentManager.java:416) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.deleteComponent(SingleComponentManager.java:163) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.doDeactivate(AbstractComponentManager.java:831) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.deactivateInternal(AbstractComponentManager.java:805) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.dispose(AbstractComponentManager.java:584) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.disposeComponents(ConfigurableComponentHolder.java:716) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.BundleComponentActivator.dispose(BundleComponentActivator.java:510) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.Activator.disposeComponents(Activator.java:429) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.Activator.access$300(Activator.java:49) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.Activator$ScrExtension.destroy(Activator.java:292) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.AbstractExtender$1.run(AbstractExtender.java:216) [42:org.apache.felix.scr:2.1.16]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at org.apache.felix.scr.impl.AbstractExtender.destroyExtension(AbstractExtender.java:238) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.AbstractExtender.bundleChanged(AbstractExtender.java:132) [42:org.apache.felix.scr:2.1.16]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:908) [?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230) [?:?]
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:213) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:120) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:112) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:168) [?:?]
	at org.eclipse.osgi.container.Module.publishEvent(Module.java:476) [?:?]
	at org.eclipse.osgi.container.Module.doStop(Module.java:634) [?:?]
	at org.eclipse.osgi.container.Module.stop(Module.java:498) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.decStartLevel(ModuleContainer.java:1675) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1594) [?:?]
	at org.eclipse.osgi.container.SystemModule.stopWorker(SystemModule.java:270) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule.stopWorker(EquinoxBundle.java:147) [?:?]
	at org.eclipse.osgi.container.Module.doStop(Module.java:636) [?:?]
	at org.eclipse.osgi.container.Module.stop(Module.java:498) [?:?]
	at org.eclipse.osgi.container.SystemModule.stop(SystemModule.java:202) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule$1.run(EquinoxBundle.java:165) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

**Comment by maggu2810** [comment]
This PR has been an initial migration and contains "some" improvements.
Yesterday I read a little bit more of the old publisher code base and realized it should be rewritten completely.
There are still to much service trackers and calls to `getService`.

I will move it all to DS if I find some time for it.

**Comment by rkoshak** [comment]
Posting this comment here before I open a new issue. Scott seems to think my problem might be related. 

On recent snapshots (build 1635) the REST API Docs will no longer load complaining about an inability to find swagger.json.

The error on the UI states:

`Can't read swagger JSON from https://<address of OH>/rest/swagger.json`

When I try to load that file directly I get:

` {"error":{"message":"HTTP 404 Not Found","http-code":404,"exception":{"class":"javax.ws.rs.NotFoundException","message":"HTTP 404 Not Found","localized-message":"HTTP 404 Not Found"}}}`

I [opened an issue over on the webuis repo](https://github.com/openhab/openhab-webui/issues/86) but it looks like to root of the problem is here. It has been reported by at least two other users on the forum.

**Comment by kaikreuzer** [comment]
@rkoshak It might be better to open a new issue here as comments on PRs that are already merged might easily be overlooked. Thanks!

**Pull request merged**